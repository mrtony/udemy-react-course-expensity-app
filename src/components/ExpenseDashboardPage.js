import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenstListFilters';

export default () => (
    <div>
      <h1>This is expense dashboard page!</h1>
      <ExpenseListFilters />
      <ExpenseList />
    </div>
);