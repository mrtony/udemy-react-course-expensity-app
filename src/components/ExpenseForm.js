import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends React.Component {
    constructor(props) {
      super(props);

      this.state = {
        description: props.expense? props.expense.description: '',
        note: props.expense? props.expense.note: '',
        amount: props.expense? (props.expense.amount / 100).toString(): '',
        createdAt: props.expense? moment(props.expense.createdAt): moment(),
        calenderFocused: false,
        error: ''
      }
    }

    onDescriptionChange = (e) => {
      const { value: description} = e.target;
      this.setState(() => ({description}));
    }
    onNoteChange = e => {
      const { value: note} = e.target;
      this.setState(() => ({note}));
    }
    onAmountChange = e => {
      const { value: amount } = e.target;
      if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/))
        this.setState(() => ({amount}));
    }
    onDateChange = (createdAt) => {
      if (createdAt)
        this.setState(() => ({createdAt}));
    }
    onFocusChange = ({focused: calenderFocused}) => {
      this.setState(() => ({calenderFocused}));
    }
    onSubmit = e => {
      e.preventDefault();
      if (!this.state.description || !this.state.amount)
        this.setState(() => ({error: 'Please provide description and amount'}));
      else {
        this.setState(() => ({error: ''}));
        this.props.onSubmit({
          description: this.state.description,
          amount: parseFloat(this.state.amount, 10) * 100,
          createdAt: +this.state.createdAt,
          note: this.state.note
        });
      }
    }
    render() {
        return (
            <div>
                <h1>Expense Form</h1>
                {
                  this.state.error && <p>{this.state.error}</p>
                }
                <form onSubmit={this.onSubmit}>
                  <input type="text" placeholder="Description" autoFocus 
                  value={this.state.description}
                  onChange={this.onDescriptionChange} />
                  <input type="text" 
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                    placeholder="Amount"/>
                  <SingleDatePicker 
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calenderFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    id="d001"
                  />
                  <textarea placeholder="Add a note for your expense(optional)"
                  value={this.state.note}
                  onChange={this.onNoteChange}>
                  </textarea>
                  <button>Add Expense</button>
                </form>
            </div>
        );
    }
}
export default ExpenseForm;