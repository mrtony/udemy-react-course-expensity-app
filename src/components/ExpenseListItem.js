import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = props => (
    <div>
        <Link to={`/edit/${props.id}`}><h3>Description: {props.description}</h3></Link>
        <p>Note: {props.amount}-{props.createdAt}</p>
    </div>
);

export default ExpenseListItem;