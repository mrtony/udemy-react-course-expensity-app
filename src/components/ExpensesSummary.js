import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectedExpenses from '../selectors/expenses';
import selectedExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({expenseCount, expensesTotal}) => (
  <div>
    <h3>Expenses Summary</h3>
    <p>Viewing {expenseCount > 1? 'expenses': 'expense'} {expenseCount} totaling {numeral(expensesTotal).format('$0,0.00')}</p>
  </div>
);

const mapStateToProps = state => {
  const visibleExpenses = selectedExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectedExpensesTotal(visibleExpenses)
  }
}

export default connect(mapStateToProps)(ExpensesSummary);