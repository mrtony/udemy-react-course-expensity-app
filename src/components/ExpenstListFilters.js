import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker, SingleDatePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import moment from 'moment';

export class ExpenseListFilters extends React.Component {
    state = {
        focusedInput: null
    };
    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }
    onFocusChange = (focusedInput) => {     
        this.setState(() => ({focusedInput}));
    }
    onTextChange = (e) => {
      this.props.setTextFilter(e.target.value);
    }
    onSortChange = (e) => {
      e.target.value === 'date'? 
      this.props.sortByDate() : this.props.sortByAmount();
    }
    render() {
        return (
            <div>
                <input type="text" 
                value={this.props.filters.text} 
                onChange={this.onTextChange}
                />
                <select 
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}
                    >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                startDateId="your_unique_start_date_id"
                startDate={this.props.filters.startDate}
                endDateId="your_unique_end_date_id"
                endDate={this.props.filters.endDate}
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.focusedInput}
                onFocusChange={this.onFocusChange}
                readOnly={true}
                isOutsideRange={() => false}
                />
            </div>
        );
    }
}

const mapStateToProps = state => ({
    filters: state.filters
});

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
