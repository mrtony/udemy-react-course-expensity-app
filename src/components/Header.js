import React from 'react';
import { connect } from 'react-redux';
import { NavLink } from 'react-router-dom';
import { startLogout } from '../actions/auth';

export const Header = ({startLogout}) => (
    <div>
        <h1>Expensity App</h1>
        <div>
        <NavLink to="/" activeClassName="is-active" exact={true}>Dashboard</NavLink>
        <NavLink to="/create" activeClassName="is-active">Add</NavLink>
        <NavLink to="/help" activeClassName="is-active">Help</NavLink>
        </div>
        <button onClick={startLogout}>Logout</button>
    </div>
);

const mapDispatchToProps = dispatch => (
  {
    startLogout: () => dispatch(startLogout())
  }
);

export default connect(undefined, mapDispatchToProps)(Header);