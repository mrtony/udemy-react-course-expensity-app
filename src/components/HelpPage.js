import React from 'react';

export default () => (
    <div>
        <h1>This is help page!</h1>
    </div>
);