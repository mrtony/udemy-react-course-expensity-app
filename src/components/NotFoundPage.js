import React from 'react';
import { Link } from 'react-router-dom';

const NotFoundPage = () => (
    <div>
        <h1>Page Not Found!!</h1>
        <Link to="/">Go Home</Link>
    </div>
);

export default NotFoundPage;