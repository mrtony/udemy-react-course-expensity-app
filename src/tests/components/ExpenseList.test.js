import React from 'react';
import moment from 'moment';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';

const expenses = [{
  id: '1',
  description: 'Gum',
  note: '',
  amount: 195,
  createdAt: moment(0)
}];

describe('ExpenseList component', () => {
  test('傳入只有1個expense的expenses陣列', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('傳入空的expenses陣列', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
  });
})