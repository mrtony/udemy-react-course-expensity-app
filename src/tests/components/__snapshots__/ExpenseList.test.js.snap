// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseList component 傳入只有1個expense的expenses陣列 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <ExpenseListItem
    amount={195}
    createdAt={"1970-01-01T00:00:00.000Z"}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
</div>
`;

exports[`ExpenseList component 傳入空的expenses陣列 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <p>
    無資料
  </p>
</div>
`;
